lib/dot.js

```diff
@@ -1,166 +1,123 @@
 /**
- * Dot density grid, using the methodology from the
- * University of Viginia Cooper Center for Public Service's Racial Dot Map,
- * http://demographics.coopercenter.org/DotMap/
+ * Dot density grid, with bilinear interpolation between opportunity density cells and multiple opacity layers
+ * for overlappig dots.
  */
 
 import {TileLayer} from 'leaflet'
 import Color from 'color'
-import MersenneTwister from 'mersenne-twister'
 
-// Dot transparency by zoom level, shamelessly ripped off from https://github.com/unorthodox123/RacialDotMap/blob/master/dotmap.pde
-const TRANSPARENCY_BY_ZOOM = {
-  0: 153,
-  1: 153,
-  2: 153,
-  3: 153,
-  4: 153,
-  5: 153,
-  6: 179,
-  7: 179,
-  8: 204,
-  9: 204,
-  10: 230,
-  11: 230,
-  12: 255,
-  13: 255,
-  14: 255,
-  15: 255,
-  16: 255,
-  17: 255,
-  18: 255,
-  19: 255,
-  20: 255,
-  21: 255,
-  22: 255
+const ZSCALE_BY_ZOOM = {
+  10: 4,
+  11: 4,
+  12: 8,
+  13: 8,
+  14: 16,
+  15: 16,
+  16: 32,
+  17: 32,
+  18: 32
 }
 
 export default class Dot extends TileLayer.Canvas {
-  initialize (grid, color = '#49a0d7', placement = Dot.halton({})) {
-    // Placement is not a lookup table of dot positions, it's a function that returns dot positions.
-    // This allows swapping in different dot positioning strategies.
-    this.placement = placement
+    
+  // grid is the grid of opportunity densities to visualize as map tiles (class from Browsochrones). 
+  // color is the color of the dots to draw.
+  initialize (grid, color = '#49a0d7') {
     this.color = new Color(color)
-    // The grid of opportunity densities to visualize as map tiles.
     this.grid = grid
   }
 
   // Given a web Mercator x, y, and zoom values for a single tile, as well as an associated canvas object,
   // visualize the opportunities falling within that tile as dots on the canvas.
-  drawTile (canvas, { x, y }, zoom) {
+  drawTile (canvas, mercTileCoord, zoom) {
     const ctx = canvas.getContext('2d')
     const imageData = ctx.createImageData(canvas.width, canvas.height)
 
     // Convert web Mercator tile position to pixels relative to the left and the top of the world at its zoom level.
-    x *= 256
-    y *= 256
-
-    const transparency = TRANSPARENCY_BY_ZOOM[zoom]
-    const rgba = [...this.color.rgbArray(), transparency]
+    const mercPixelX = mercTileCoord.x * 256
```

Comment:
True, I thought about that earlier but didn't know the details of how Leaflet handles retina tiles.